File: Data Structures & Algorithms/longest-increasing-path-in-matrix/submission-9.py
class Solution:
    def longestIncreasingPath(self, matrix: List[List[int]]) -> int:
        rows = len(matrix) 
        cols = len(matrix[0]) 

        dp = [ [ -1 for _ in range(cols) ] for _ in range(rows) ] 

        def dfs(r, c): 
            if dp[r][c] >= 0: 
                return dp[r][c] 

            best = 1 
            for dr, dc in [(1, 0), (0, 1), (-1, 0), (0, -1)]: 
                nr, nc = r + dr, c + dc 
                if 0 <= nr < rows and 0 <= nc < cols and matrix[r][c] < matrix[nr][nc]: 
                    best = max(best, 1 + dfs(nr, nc)) 

            dp[r][c] = best 
            return best 

        for r in range(rows): 
            for c in range(cols): 
                dfs(r, c) 

        res = 0 
        for r in range(rows): 
            for c in range(cols): 
                res = max(res, dp[r][c]) 

        return res

File: Data Structures & Algorithms/search-2d-matrix/submission-9.js
class Solution {
    /**
     * @param {number[][]} matrix
     * @param {number} target
     * @return {boolean}
     */
    searchMatrix(matrix, target) { 

        let rows = matrix.length 
        let cols = matrix[0].length 

        let l = 0 
        let r = rows * cols - 1 

        while(l <= r) {
            let mid = Math.floor( ( l + r ) / 2 ) 
            let midVal = matrix[ Math.floor( mid / cols ) ][ mid % cols ] 

            if (midVal === target) {
                return true
            } else if (midVal < target) {
                l = mid + 1
            } else {
                r = mid - 1
            }
        }     

        return false
    }
}
